build/images/scripts/install_ovs_windows.ps1

```diff
@@ -0,0 +1,131 @@
+Param(
+    [parameter(Mandatory = $false)] [string] $DownloadDir,
+    [parameter(Mandatory = $false)] [string] $DownloadURL,
+    [parameter(Mandatory = $false)] [string] $OVSInstallDir = "C:\openvswitch"
+)
+
+$ErrorActionPreference = "Stop"
+$OVSDownloadURL = "https://github.com/ruicao93/ovs/releases/download/v2.12.9-win64/ovs-2.12.90-win64.zip"
+$OVSDownloadDir = [System.IO.Path]::GetDirectoryName($myInvocation.MyCommand.Definition)
+$InstallLog = "$OVSDownloadDir\install.log"
+$OVSZip = "$OVSDownloadDir\ovs-win64.zip"
+
+if ($DownloadDir -ne "") {
+    $OVSDownloadDir = $DownloadDir
+    $InstallLog = "$OVSDownloadDir\install.log"
+}
+
+if ($DownloadURL -ne "") {
+    $OVSDownloadURL = $DownloadURL
+}
+
+function Log($Info) {
+    $time = $(get-date -Format g)
+    "$time $Info `n`r" | Tee-Object $InstallLog -Append
+}
+
+function CreatePath($Path){
+    if ($(Test-Path $Path)) {
+        mv $Path $($Path + "_bak")
+    }
+    mkdir -p $Path | Out-Null
+}
+
+function SetEnvVar($key, $value) {
+    [Environment]::SetEnvironmentVariable($key, $value, [EnvironmentVariableTarget]::Machine)
+}
+
+function WaitExpandFiles($Src, $Dest) {
+    Log "Extract $Src to $Dest"
+    Expand-Archive -Path $Src -DestinationPath $Dest | Out-Null
+}
+
+function CheckIfOVSInstalled() {
+    If (Get-Service ovs-vswitchd -ErrorAction SilentlyContinue) {
+        Log "Found existing OVS service, exit OVS installation."
+        exit 0
+    }
+}
+
+function DownloadOVS() {
+    if (!(Test-Path $OVSDownloadDir)) {
+        mkdir -p $OVSDownloadDir
+    }
+    Log "Downloading OVS package from $OVSDownloadURL to $OVSZip"
+    curl.exe -sLo $OVSZip $OVSDownloadURL
+    if (!$?) {
+        Log "Download OVS failed, URL: $OVSDownloadURL"
+        exit 1
+    }
+    Log "Download OVS package success."
+}
+
+function InstallOVS() {
+    # unzip OVS.
+    WaitExpandFiles $OVSZip $OVSDownloadDir
+    # Copy OVS package to target dir.
+    Log "Copying OVS package from $OVSDownloadDir\openvswitch to $OVSInstallDir"
+    cp -r "$OVSDownloadDir\openvswitch" $OVSInstallDir
+    # Create log and run dir.
+    $OVS_LOG_PATH = $OVSInstallDir + "\var\log\openvswitch"
+    CreatePath $OVS_LOG_PATH
+    $OVSRunDir = $OVSInstallDir + "\var\run\openvswitch"
+    CreatePath $OVSRunDir
+    # Install OVS driver certificate.
+    Log "Installing OVS driver certificate."
+    $OVSDriverDir = "$OVSInstallDir\driver"
+    $cert = New-Object System.Security.Cryptography.X509Certificates.X509Certificate2("$OVSDriverDir\package.cer")
```

Comment:
I'm trying to understand what's going on here. I think it is related to what we discussed yesterday.

`driver\package.cer` a self-signed certificate that you used to sign the OVS binaries. Are there any requirements for adding this certificate to the trust store (yesterday you said something about "test mode"), or will Windows just prompt for confirmation?

@McCodeman

Reply:
The reason we add the cert to trust store is to avoid windows prompt for confirmation.

And actually there's no requirement for adding the cert. The "test mode" is required if user use unsigned OVS driver. For the details of the requirements needed by OVS can be found in `The Kernel Module` section of OVS doc:
http://docs.openvswitch.org/en/latest/intro/install/windows/

Reply:
Thanks for the link.

So just to confirm, there are 2 cases:
 * when running this script, the user will add our self-signed certificate (included in the archive) to the trust store. The `TESTSIGNING` boot option is not required.
 * if the user builds their own version of OVS (by following instructions in http://docs.openvswitch.org/en/latest/intro/install/windows/ and do not sign the binaries, then they need `TESTSIGNING`.

I want to check with you because I am confused by your statement "And actually there's no requirement for adding the cert". It seems to me that adding the certificate is a requirement unless I want to enable `TESTSIGNING`.

Reply:
Sorry Antonin, I think my expain is not accurate.

Please forget what I have said, and let me make it clear:
- Commonly,  a kernel driver need to be [signed by Windows Hardware Dev Center](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/driver-signing) then it can be loaded.
- For development, we can `test-sign` the driver for test. The OVS package we provided now is test-signed. If user want to used the `test-signed` OVS driver, they need to do both following works per [windows doc](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/installing-a-test-signed-driver-package-on-the-test-computer):
  - Set `TESTSIGNING` as on.
  - Copy test certificate is copied to the Trusted Root Certification Authorities certificate store.